Add a Swagger UI extension

danielpetisme opened the pull request:
Proposal to fix #1355

The extension declares a SevletExtension to intercept the request and delegates the Swagger UI resolution to a dedicated ResourceHandler (as discussed in the issue).

The ServletExtensionBuildItem is produced only during the dev/test launch mode. The swagger-ui path is configurable via the `quarkus.swagger-ui.path` property.

The swagger-ui is bundled with the *deployment* artifact. Maven downloads the Web assets during the `prepare-resources` phase, replaces the sample Open API URL location by `/openapi/` to match smallrye-openapi extension before copying it into the `target/classes/swagger-ui` directory. That means there is not any sort of customization of the swagger-ui front (html, js or css update) and the maintenance is limited to update the swagger-ui webjar version (managed as a usual maven dependency).

danielpetisme committed:
Introduce a Swagger UI extension

danielpetisme commented:
@kenfinnigan @gsmet could you review it 🙏

gsmet commented on extensions/swagger-ui/deployment/src/main/java/io/quarkus/swaggerui/deployment/SwaggerUiProcessor.java:
```diff
@@ -0,0 +1,68 @@
+     */
+    @BuildStep
+    FeatureBuildItem feature() {
+        log.info("Feature: swagger-ui");
```
This needs to be removed.

danielpetisme replied:
Done

gsmet commented on extensions/swagger-ui/deployment/src/main/java/io/quarkus/swaggerui/deployment/SwaggerUiProcessor.java:
```diff
@@ -0,0 +1,68 @@
+    @BuildStep
+    FeatureBuildItem feature() {
+        log.info("Feature: swagger-ui");
+        return new FeatureBuildItem("swagger-ui");
```
This should be added as a constant in `FeatureBuildItem`. At least, that's the convention for now. Please follow the alphabetical order there.

danielpetisme replied:
Log removed + Constant introduced

gsmet commented on extensions/swagger-ui/deployment/src/main/java/io/quarkus/swaggerui/deployment/SwaggerUiProcessor.java:
```diff
@@ -0,0 +1,68 @@
+    public void registerSwaggerUiServletExtension(SwaggerUiTemplate template,
+            BuildProducer<ServletExtensionBuildItem> servletExtension, BeanContainerBuildItem container) {
+        if (launch.getLaunchMode().isDevOrTest()) {
+            log.info("Registering Swagger UI Servlet Extension");
```
Please remove that. We try to reduce the log noise to a minimum.

danielpetisme replied:
Log removed

gsmet commented on extensions/swagger-ui/pom.xml:
```diff
@@ -0,0 +1,26 @@
+    <packaging>pom</packaging>
+
+    <properties>
+        <swagger-ui.version>3.20.9</swagger-ui.version>
```
This needs to be added to the BOM (together with the dependency declaration).

danielpetisme replied:
I've an issue between the `bom` and `build-parent` management.
The `bom` defines the dependencies but I can inherit a property. I can do it with the `build-parent.
I need this property for a path construction
https://github.com/quarkusio/quarkus/pull/1573/files#diff-021ff459d881cd8da34e1d20241adb8cR77

ATM, I've the `swagger-ui.version` version in both poms... Do you see any proper way to do that?

gsmet replied:
Hmmm, I missed that. I checked both plugins for which you used the version and I think you could make it work with an `*` as some of the directives of both plugins support that.

At least I think it's worth a try.

danielpetisme replied:
:thinking:  not sure to understand your comment? You mean define the dependency version in the `bom` and remove the property from the `build-parent` + replace the replace `${swaggerui.version}` by `*` like `${project.build.directory}/webjar-swagger-ui/META-INF/resources/webjars/swagger-ui/*/` ?

gsmet replied:
Yes, exactly. I think it could work as at least some of the properties support `*`, for both plugins.

danielpetisme replied:
With @stuartwdouglas proposal, we don't need this anymore.

gsmet commented on extensions/swagger-ui/runtime/pom.xml:
```diff
@@ -0,0 +1,128 @@
+            <plugin>
+                <groupId>com.google.code.maven-replacer-plugin</groupId>
+                <artifactId>replacer</artifactId>
+                <version>${replacer.version}</version>
```
The plugin and the version should be declared in `build-parent/pom.xml` together with the others.

Note that for now, we don't have properties for the versions of the plugins so please keep it consistent with the others in `build-parent`.

danielpetisme replied:
Done, but I would appreciate a review on that part to ensure I'm aligned with Quarkus policies.

gsmet replied:
Looks good.

gsmet commented on extensions/swagger-ui/runtime/pom.xml:
```diff
@@ -0,0 +1,128 @@
+            </plugin>
+            <plugin>
+                <artifactId>maven-resources-plugin</artifactId>
+                <version>3.1.0</version>
```
Same here about having the version declaration in `build-parent`.

gsmet commented on extensions/swagger-ui/runtime/pom.xml:
```diff
@@ -0,0 +1,128 @@
+                            <goal>copy-resources</goal>
+                        </goals>
+                        <configuration>
+                            <outputDirectory>${project.build.directory}/classes/swagger-ui</outputDirectory>
```
I'm not totally sold by the location of these files.

@stuartwdouglas do you have a preference on where we should put web resources that we don't want publicly accessible?

stuartwdouglas replied:
Something I don't really like about this is that in JVM mode these are still copied into the uberjar or lib dir

stuartwdouglas replied:
So I think an alternative approach would be to:
- Add the webjar as a dependency of the deployment module
- In dev mode as part of the build step extract this to a temp directory (and store the result in a static)
- Register the result as a special resource handler

This means that none of the swagger-ui ends up in the runtime artifact

danielpetisme replied:
Does it make sense to create a `${cwd}/.quarkus` directory, copy the swagger-ui into and serve it?

stuartwdouglas replied:
I think it would make more sense to just put it in a temp dir, end remove it with a shutdown hook

danielpetisme replied:
@stuartwdouglas done.
I propose to close this conversation for readability sake and continue the discussion on this diff
https://github.com/quarkusio/quarkus/pull/1573/files#diff-3d22741acca0b68514ba03801f7818d3

gsmet reviewed (changes requested):
@danielpetisme I like the use of a webjar and the replacement plugin. Very clever.

I added a few minor comments. I'll test it later today with an example.

Would you be interested in writing a guide/quickstart about how to use OpenAPI and Swagger UI?

Something like https://github.com/quarkusio/quarkus/blob/master/docs/src/main/asciidoc/fault-tolerance-guide.adoc (and the associated quickstart).

You would have to play a bit with OpenAPI and I think it could be an interesting task for you.

gsmet commented on extensions/swagger-ui/deployment/src/test/java/io/quarkus/swaggerui/deployment/NoConfigTest.java:
```diff
@@ -0,0 +1,26 @@
+    @Test
+    public void shouldUseDefaultConfig() {
+        RestAssured.when().get("/swagger-ui").then().statusCode(200);
+        RestAssured.when().get("/swagger-ui/index.html").then().statusCode(200);
```
One additional comment: I wonder if we should test if the pattern `/openapi` is in the page returned. IIRC, it would be something like `.body(containsString("/openapi"))`.

danielpetisme replied:
Done
https://github.com/quarkusio/quarkus/pull/1573/files#diff-28ec27bba9df0f01afcda9af942bdc63R22

kenfinnigan commented:
I think we'd discussed having this extension automatically added if `quarkus-smallrye-openapi` was added. I don't see such a dependency included in this PR.

One other thing, to not clutter URL space, I'm wondering if the URL for the UI should be `/openapi/ui` instead?

danielpetisme commented:
@gsmet yes I can propose a guide too/quickstart.

@kenfinnigan Regarding the swagger-ui/openapi coupling. I used to prefer explicit imports. IMHO, one can use openapi and not forced  to use swagger-ui. One scenario I can see is to customizer swagger-ui (update the CSS, the version or enable it for production under a authenticated path). So far the PR don't propose advanced configuration options (IMHO, we don't have too) so a dev should be able to use open api and import its own swagger-ui into its project as I did in my first prototype.

Regarding the path, `/swagger-ui` is quite standard (or at least on the project I worked on). `openapi/ui` loose the information of having explicitly `swagger-ui` embedded. Its the dev responsability to update it if needed (actually you can set `quarkus.swagger-ui.path=/openapi/ui) if you want to

kenfinnigan commented:
@danielpetisme I recall we discussed on the issue that the Swagger UI should just "be there" in dev mode when `quarkus-smallrye-openapi` is included, without needing to explicitly add it.

My concern with the URL is we don't want to pollute the URL namespace with too many paths. 1) It increases the chances of clashing with something from the user, which is more complicated when it's only there for dev mode. 2) It means more things need to be secured if needed (Related to this we may consider putting all endpoints like this under `/admin` to make that easier)

danielpetisme commented:
Let's say I swagger-ui and openapi are bundled together, I create a project and I want to add a modified version of swagger-ui, I'll copy the swagger-ui directory into my project resources directory. Considering swagger-ui already bundled with openapi, how quarkus should behaves when the user goes to `/swagger-ui`? By an explicit import the dev is aware that this path is used.

Another proposal would be to be to keep `/dev/*` for the dev/test oriented extensions ?

gsmet commented:
Just so that we don't forget about it: we discussed with Daniel and he hasn't tested native image support yet. There's a good chance we will need a bunch of `SubstrateResourceBuildItem`.

danielpetisme commented:
I've implemented @stuartwdouglas proposal.
During deployment I find the swagger artifact thanks to the built-in `ClassPathArtifactResolver`. Then I copy the jar's content into a temp directory (+ openapi url update). Finally, I provide the HTTP path and the temp dir path to the `SwaggerUiTemplate` to build a Resource Handler with these 2 piece of information.

I've added a Shutdown Context in order to clean the temp dir. I don't now how I can test it programmatically (since it's a mix of deployment and runtime) but after some manual tests it seems to work properly.

I'm wondering if it makes sense to exposes the SwaggerUI resource dir as a configuration item (if empty, use `java.io.tmpdir`) and maybe the Open API url too (default to `/openapi`).
WDYT?

danielpetisme committed:
Add a guide for the Swagger UI extension

danielpetisme commented:
@gsmet @kenfinnigan @stuartwdouglas I've started the guide/quickstart too. 🙏 if you have some time to have an eye on them too.

gsmet commented on extensions/swagger-ui/deployment/src/main/java/io/quarkus/swaggerui/deployment/SwaggerUiProcessor.java:
```diff
@@ -72,6 +72,7 @@ public void registerSwaggerUiServletExtension(SwaggerUiTemplate template,
             BuildProducer<ServletExtensionBuildItem> servletExtension,
             BeanContainerBuildItem container,
             ShutdownContextBuildItem shutdown) {
+        log.warn("Here");
```
You don't want that one in.

danielpetisme replied:
Oops, we'll clean.

gsmet commented on extensions/swagger-ui/runtime/src/main/java/io/quarkus/swaggerui/runtime/SwaggerUiServletExtension.java:
```diff
@@ -47,10 +47,9 @@ public void setResourceDir(String resourceDir) {
-        //                SWAGGER_UI_RESOURCES_DIR);
         deploymentInfo.addOuterHandlerChainWrapper(
                 (handler) -> Handlers.path(handler).addPrefixPath(path, resource(resourceManager)));
+        log.info("Swagger UI served at " + path);
```
I don't think we want that one either. At least, not at info level. IIRC, you had a debug level message similar to this one somewhere?

danielpetisme replied:
Normally, I removed the other logs. I only keep this one, for me it makes sense to display this info to remember the default path/customized path.
```
11:29:20,129 INFO  [io.qua.dep.QuarkusAugmentor] Beginning quarkus augmentation
11:29:20,337 INFO  [org.jbo.threads] JBoss Threads version 3.0.0.Alpha4
11:29:20,473 WARN  [io.qua.config] Unrecognized configuration key "quarkus.undertow.resources" provided
11:29:20,766 WARN  [io.qua.swa.dep.SwaggerUiProcessor] Here
11:29:21,002 INFO  [io.qua.dep.QuarkusAugmentor] Quarkus augmentation completed in 873ms
11:29:21,031 INFO  [io.sma.ope.api.OpenApiDocument] OpenAPI document initialized: io.smallrye.openapi.api.models.OpenAPIImpl@4a003cbe
11:29:21,078 INFO  [io.qua.swa.run.SwaggerUiServletExtension] Swagger UI accessible at /swagger-ui
11:29:21,246 INFO  [org.xnio] XNIO version 3.7.0.Final
11:29:21,263 INFO  [org.xni.nio] XNIO NIO Implementation Version 3.7.0.Final
11:29:21,368 INFO  [io.quarkus] Quarkus 999-SNAPSHOT started in 1.371s. Listening on: http://[::]:8080
11:29:21,368 INFO  [io.quarkus] Installed features: [cdi, resteasy, smallrye-openapi, swagger-ui]
```

gsmet reviewed (changes requested):
@danielpetisme I just took a quick look at some of the latest fixes, I will do a proper review later. Could you also squash things a bit to have more meaningful commits?

One for the extension and one for the documentation would seem like a nice thing.

danielpetisme commented:
Squash done

gsmet commented on docs/src/main/asciidoc/openapi-swaggerui-guide.adoc:
```diff
@@ -0,0 +1,249 @@
+}
+----
+
+Notice that there is no API specification or presentation specific code
```
Maybe add `yet.` at the end? The final dot is missing in any case.

gsmet commented on docs/src/main/asciidoc/openapi-swaggerui-guide.adoc:
```diff
@@ -0,0 +1,249 @@
+
+== Expose OpenAPI Specifications
+
+Quarkus proposes a `smallrye-openapi` extension compliant with https://github.com/eclipse/microprofile-open-api/[Eclipse MicroProfile OpenAPI]
```
with *the*

gsmet commented on docs/src/main/asciidoc/openapi-swaggerui-guide.adoc:
```diff
@@ -0,0 +1,249 @@
+== Expose OpenAPI Specifications
+
+Quarkus proposes a `smallrye-openapi` extension compliant with https://github.com/eclipse/microprofile-open-api/[Eclipse MicroProfile OpenAPI]
+specification in order to generate your API's https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md[OpenAPI v3 specification].
```
I would drop the `'s`

gsmet commented on docs/src/main/asciidoc/openapi-swaggerui-guide.adoc:
```diff
@@ -0,0 +1,249 @@
+== Use Swagger UI for development
+
+When building APIs, developers want to test them quickly. https://swagger.io/tools/swagger-ui/[Swagger UI] is a great tool
+permitting to visualize and interact with your APIs. It's automatically generated via your OpenAPI specification.
```
I would go with `The UI is automatically generated from your OpenAPI specification.`

gsmet commented on docs/src/main/asciidoc/openapi-swaggerui-guide.adoc:
```diff
@@ -0,0 +1,249 @@
+
+Quarkus proposes a `swagger-ui` extension embedding a Swagger UI page configured to find your API specification at `/openapi`.
+
+*Note*: Currently the swagger-ui page is available when Quarkus is started in a dev or test mode.
```
s/Currently the swagger-ui page is available/Currently, Swagger UI is only available/
s/in a dev or test mode/in dev or test mode/

gsmet commented on docs/src/main/asciidoc/openapi-swaggerui-guide.adoc:
```diff
@@ -0,0 +1,249 @@
+When building APIs, developers want to test them quickly. https://swagger.io/tools/swagger-ui/[Swagger UI] is a great tool
+permitting to visualize and interact with your APIs. It's automatically generated via your OpenAPI specification.
+
+Quarkus proposes a `swagger-ui` extension embedding a Swagger UI page configured to find your API specification at `/openapi`.
```
s/proposes/comes with/ to change a bit?
s/a Swagger UI page configured to find your API specification at `/openapi`/a properly configured Swagger UI page/ ?

gsmet commented on docs/src/main/asciidoc/openapi-swaggerui-guide.adoc:
```diff
@@ -0,0 +1,249 @@
+
+*Note*: Currently the swagger-ui page is available when Quarkus is started in a dev or test mode.
+
+The first is to add the `swagger-ui` extension to your Quarkus application:
```
s/The first/The first step/

gsmet commented on docs/src/main/asciidoc/openapi-swaggerui-guide.adoc:
```diff
@@ -0,0 +1,249 @@
+mvn quarkus:add-extension -Dextensions="swagger-ui"
+----
+
+By default, SwaggerUI is accessible at `/swagger-ui`. You can update this path with the `quarkus-swagger-ui.path` property.
```
s/SwaggerUI/Swagger UI/
s/quarkus-swagger-ui.path/quarkus.swagger-ui.path/ (with a dot)

gsmet commented on docs/src/main/asciidoc/openapi-swaggerui-guide.adoc:
```diff
@@ -0,0 +1,249 @@
+By default, SwaggerUI is accessible at `/swagger-ui`. You can update this path with the `quarkus-swagger-ui.path` property.
+Here you have an example of configuration:
+----
+quarkus-swagger-ui.path=/my-custom-path
```
dot it is.

gsmet commented on docs/src/main/asciidoc/openapi-swaggerui-guide.adoc:
```diff
@@ -0,0 +1,249 @@
+
+You can check the Swagger UI path in your application's log
+----
+00:00:00,000 INFO  [io.qua.swa.run.SwaggerUiServletExtension] Swagger UI accessible at /swagger-ui
```
Ah, now I understand why you have this log info. Considering it only appears in dev mode, maybe it makes sense to have it.

gsmet commented on docs/src/main/asciidoc/openapi-swaggerui-guide.adoc:
```diff
@@ -0,0 +1,249 @@
+00:00:00,000 INFO  [io.qua.swa.run.SwaggerUiServletExtension] Swagger UI accessible at /swagger-ui
+----
+
+Once your application is started, you can go to http://localhost:8080/swagger-ui and play with your API
```
Missing final dot.

gsmet commented on docs/src/main/asciidoc/openapi-swaggerui-guide.adoc:
```diff
@@ -0,0 +1,249 @@
+You can visualize your API's operations and schemas.
+image:openapi-swaggerui-guide-screenshot01.png[alt=Visualize your API]
+
+You also interact with your API in order to quickly test it.
```
+ can

gsmet reviewed (commented):
Here is a review of the documentation part.

I think at some point, we will want to add more information about OpenAPI and how you can add additional information through annotations but it seems to be a good first step.

danielpetisme commented:
@gsmet Should I close the outdated conversation or you do it?

stuartwdouglas commented on extensions/swagger-ui/deployment/src/main/java/io/quarkus/swaggerui/deployment/SwaggerUiProcessor.java:
```diff
@@ -0,0 +1,128 @@
+            try {
+                ResolvedArtifact artifact = getSwaggerUiArtifact();
+                Path tempDir = Files.createTempDirectory(TEMP_DIR_PREFIX);
+                extractSwaggerUi(artifact, tempDir);
```
We really only need to do this once per JVM execution. I think this should be cached in a static, and use a Shutdown hook to clean it up. We don't really want to have to do this unzip after every code change.

stuartwdouglas commented on extensions/swagger-ui/deployment/src/main/java/io/quarkus/swaggerui/deployment/SwaggerUiProcessor.java:
```diff
@@ -0,0 +1,128 @@
+
+    private void updateApiUrl(Path indexHtml) throws IOException {
+        String content = new String(Files.readAllBytes(indexHtml));
+        content = content.replaceAll(SWAGGER_UI_DEFAULT_API_URL, OPEN_API_DEFAULT_URL);
```
If this has been modified by config shouldn't this be taken into account? Also if you make both these changes then you would need to also cache this path, so if it changes you would need to update the indexHtml accordingly (or just delete and re-run).

gsmet commented on docs/src/main/asciidoc/openapi-swaggerui.adoc:
```diff
@@ -0,0 +1,45 @@
+= {project-name} - Using OpenAPI and Swagger UI
+
+This guide explains how your Quarkus application can expose its API description through an OpenAPI specification and
+how you can test it via a user-friendly Swagger UI.
```
This one hasn't been fixed: there's still a `openapi-swaggerui.adoc` document created in the first commit, containing some of the OpenTracing guide.

gsmet committed:
Add a common FileUtil class

gsmet reviewed (approved):
I made some additional changes to fix minor issues and it now looks good to go.

Thanks @stuartwdouglas for helping on the archive deployment side. @kenfinnigan due to Stuart's work, the dependencies are the other way around: the Swagger UI extension depends on the OpenAPI one as we need to inject its configuration.

gsmet commented:
Waiting for CI to go green.

kenfinnigan commented:
That's not very user friendly though is it?

It means you no longer get the Swagger UI in dev mode by just adding the OpenAPI extension.

gsmet commented:
Yes, you have to add it explicitly. It's a bit of a chicken and egg issue.

But it doesn't seem that bad as you need to be aware of it anyway.

Adding the `swagger-ui` extension will add both.

kenfinnigan commented:
Why not something similar to RESTEasy where there's an openapi-common that has the config in it?

Then both extensions depend on that, and openapi can still depend on swaggerui.

gsmet commented:
> Why not something similar to RESTEasy where there's an openapi-common that has the config in it?

We could do that.

I'll merge this PR today anyway as it's a good step forward. We can build on it early next week to have something finalized for the next release.

gsmet reviewed (dismissed):
Please don’t merge. I have to squash the fixup commits.

gsmet reviewed (approved):
Let’s merge. We will make some adjustments next week.

gsmet commented:
Merged, thanks everyone and especially @danielpetisme
